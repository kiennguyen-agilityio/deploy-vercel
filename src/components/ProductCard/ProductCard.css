.product-card {
  width: 295px;
  height: 363px;
  margin: 0 0 50px 15px;
  position: relative;
}

.product-card img {
  width: 100%;
  max-height: 300px;
  height: 259px;
  object-fit: cover;
  margin-bottom: 15px;
}

.product-card p {
  font-size: var(--font-size-regular);
  margin-bottom: 4px;
}

.product-card .sell-count {
  width: 73px;
  height: 14px;
  font: normal var(--font-weight-regular) var(--font-size-regular) / var(--line-height-regular)
    'Open Sans';
  letter-spacing: 0.1px;
  color: var(--color-price);
  margin-left: 11.5px;
}

.product-card .name {
  width: 254px;
  height: 27px;
  font: normal var(--font-weight-bold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
  color: var(--color-overlay);
  margin-bottom: 11.5px;
}

.product-card .price {
  width: 52px;
  height: 24px;
  font: normal var(--font-weight-extrabold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
  display: flex;
  align-items: center;
  color: var(--color-overlay);
}

.rating i {
  color: var(--color-icon-tertiary);
  margin-right: 4px;
}

.heart-icon {
  position: absolute;
  top: 10px;
  right: 10px;
  width: 30px;
  height: 30px;
  border-radius: 50%;
  background: rgba(255, 255, 255, 0.9);
  display: flex;
  align-items: center;
  justify-content: center;
}

.heart-icon .fa-heart {
  border: none;
  color: var(--color-primary);
}

.product-discount {
  display: flex;
  align-items: center;
}

.product-cost {
  font: normal var(--font-weight-regular) var(--font-size-regular) / var(--line-height-md)
    'Open Sans';
  color: var(--color-primary);
  margin: 0 10px 0 10px;
}

.discount-percent {
  font: normal var(--font-weight-regular) var(--font-size-regular) / var(--line-height-md)
    'Open Sans';
  color: var(--color-secondary);
}
