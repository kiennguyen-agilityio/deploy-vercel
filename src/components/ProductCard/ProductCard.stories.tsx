import { StoryObj, Meta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';

import ProductCard from './ProductCard';

export default {
  title: 'Components/ProductCard',
  component: ProductCard,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
} as Meta;

type Story = StoryObj<typeof ProductCard>;

export const Default: Story = {
  args: {
    product: {
      id: 1,
      name: 'Product Name',
      sale: '1',
      price: 100,
      discountPercent: 20,
      image: 'src/assets/images/black-chair.png',
      descriptions: 'Product descriptions',
      quantity: 5,
      color: 'Red',
      subcategory: 'Electronics',
    },
  },
};
