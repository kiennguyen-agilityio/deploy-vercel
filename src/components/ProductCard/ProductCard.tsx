import { memo } from 'react';
import { Link } from 'react-router-dom';

// interfaces
import Product from '@/interface/Product';

// style
import './ProductCard.css';

interface Props {
  product: Product;
}

const ProductCard: React.FC<Props> = ({ product }) => {
  const handleRenderRating = (rating: number) => {
    const stars = Array.from({ length: 5 }, (_, index) => (
      <i key={index} className={`fa-solid fa-star${index < rating ? ' checked' : ''}`} />
    ));

    return stars;
  };

  const handleRenderLabel = (quantity: number, discountPercent: number) => {
    if (product.quantity <= 0 || product.discountPercent > 0) {
      const discountedPrice = product.price - product.price * (discountPercent / 100);

      return (
        <div className="product-discount">
          {quantity === 0 ? (
            <span>Sold out</span>
          ) : (
            <>
              <span className="discounted-price">${discountedPrice.toFixed(0)}</span>
              <del className="product-cost">${product.price.toFixed(0)}</del>
              <span className="discount-percent">-{discountPercent}%</span>
            </>
          )}
        </div>
      );
    }

    return null;
  };

  return (
    <article className={`product-card ${product.id}`}>
      <Link to={`/products/${product.id}`} key={product.id}>
        <img src={product.image} alt={product.name} />
        <h3 className="name">{product.name}</h3>
        <div className="rating">
          {handleRenderRating(5)}
          <span className="sell-count">({product.sale} sells)</span>
        </div>
        <span className="price">
          {product.discountPercent > 0
            ? handleRenderLabel(product.quantity, product.discountPercent)
            : `$${product.price.toFixed(0)}`}
        </span>

        <div className="heart-icon">
          <i className="fa-solid fa-heart" />
        </div>
      </Link>
    </article>
  );
};

export default memo(ProductCard);
