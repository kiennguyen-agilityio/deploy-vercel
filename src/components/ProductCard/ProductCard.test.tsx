import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import ProductCard from './ProductCard';

const product = {
  id: 1,
  name: 'Fullset Black Chair & Sofa',
  sale: '350',
  price: 5000,
  discountPercent: 10,
  image: '../src/assets/images/full-black-chair-sofa.png',
  descriptions:
    'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
  quantity: 2,
  color: 'Red',
  subcategory: 'Office chair',
};

const setup = (props = {}) => {
  return render(
    <MemoryRouter>
      <ProductCard product={product} {...props} />
    </MemoryRouter>
  );
};

describe('ProductCard component', () => {
  it('renders ProductCard component correctly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('renders product information correctly', () => {
    const { getByText } = setup();
    expect(getByText(product.name)).toBeInTheDocument();
    expect(getByText(`(${product.sale} sells)`)).toBeInTheDocument();
    expect(getByText(`$${product.price.toFixed(0)}`)).toBeInTheDocument();
  });

  it('renders correct number of stars based on rating', () => {
    const { container } = setup();
    const ratingContainer = container.querySelector('.rating');
    expect(ratingContainer).toMatchSnapshot();
  });

  it('renders correct label when has discount', () => {
    const { getByText, queryByText } = setup();

    const hasDiscount = product.discountPercent > 0;
    const discountedPrice = product.price - product.price * (product.discountPercent / 100);

    expect(
      getByText(`$${hasDiscount ? discountedPrice.toFixed(0) : product.price.toFixed(0)}`)
    ).toBeInTheDocument();

    expect(getByText(`-${product.discountPercent}%`)).toBeInTheDocument();
    expect(queryByText('Sold out')).toBeNull();
  });

  it('renders correct link to product detail page', () => {
    const { getByText } = setup();
    const link = getByText(product.name).closest('a');

    expect(link).toHaveAttribute('href', `/products/${product.id}`);
  });
});
