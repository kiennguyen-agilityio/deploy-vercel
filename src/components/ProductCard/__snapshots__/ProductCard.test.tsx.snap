// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component renders ProductCard component correctly 1`] = `
<div>
  <article
    class="product-card 1"
  >
    <a
      href="/products/1"
    >
      <img
        alt="Fullset Black Chair & Sofa"
        src="../src/assets/images/full-black-chair-sofa.png"
      />
      <h3
        class="name"
      >
        Fullset Black Chair & Sofa
      </h3>
      <div
        class="rating"
      >
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <span
          class="sell-count"
        >
          (
          350
           sells)
        </span>
      </div>
      <span
        class="price"
      >
        <div
          class="product-discount"
        >
          <span
            class="discounted-price"
          >
            $
            4500
          </span>
          <del
            class="product-cost"
          >
            $
            5000
          </del>
          <span
            class="discount-percent"
          >
            -
            10
            %
          </span>
        </div>
      </span>
      <div
        class="heart-icon"
      >
        <i
          class="fa-solid fa-heart"
        />
      </div>
    </a>
  </article>
</div>
`;

exports[`ProductCard component renders correct number of stars based on rating 1`] = `
<div
  class="rating"
>
  <i
    class="fa-solid fa-star checked"
  />
  <i
    class="fa-solid fa-star checked"
  />
  <i
    class="fa-solid fa-star checked"
  />
  <i
    class="fa-solid fa-star checked"
  />
  <i
    class="fa-solid fa-star checked"
  />
  <span
    class="sell-count"
  >
    (
    350
     sells)
  </span>
</div>
`;

exports[`renders ProductCard component correctly 1`] = `
<div>
  <article
    class="product-card 1"
  >
    <a
      href="/products/1"
    >
      <img
        alt="Fullset Black Chair & Sofa"
        src="../src/assets/images/full-black-chair-sofa.png"
      />
      <h3
        class="name"
      >
        Fullset Black Chair & Sofa
      </h3>
      <div
        class="rating"
      >
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <i
          class="fa-solid fa-star checked"
        />
        <span
          class="sell-count"
        >
          (
          350
           sells)
        </span>
      </div>
      <span
        class="price"
      >
        <div
          class="product-discount"
        >
          <span
            class="discounted-price"
          >
            $
            4500
          </span>
          <del
            class="product-cost"
          >
            $
            5000
          </del>
          <span
            class="discount-percent"
          >
            -
            10
            %
          </span>
        </div>
      </span>
      <div
        class="heart-icon"
      >
        <i
          class="fa-solid fa-heart"
        />
      </div>
    </a>
  </article>
</div>
`;
